gadget/gadget.go

```diff
@@ -299,41 +329,106 @@ func validateVolume(name string, vol *Volume) error {
 		return fmt.Errorf("invalid schema %q", vol.Schema)
 	}
 
-	structureNames := make(map[string]bool, len(vol.Structure))
+	// named structures, for cross-referencing relative offset-write names
+	knownStructures := make(map[string]*PositionedStructure, len(vol.Structure))
+	// for validating structure overlap
+	structures := make([]PositionedStructure, len(vol.Structure))
+
+	previousEnd := Size(0)
 	for idx, s := range vol.Structure {
 		if err := validateVolumeStructure(&s, vol); err != nil {
-			return fmt.Errorf("invalid structure %s: %v", fmtIndexAndName(idx, s.Name), err)
+			return fmt.Errorf("invalid structure %v: %v", fmtIndexAndName(idx, s.Name), err)
 		}
+		var start Size
+		if s.Offset != nil {
+			start = *s.Offset
+		} else {
+			start = previousEnd
+		}
+		end := start + s.Size
+		ps := PositionedStructure{
+			VolumeStructure: &vol.Structure[idx],
+			StartOffset:     start,
+			index:           idx,
+		}
+		structures[idx] = ps
 		if s.Name != "" {
-			if structureNames[s.Name] {
+			if _, ok := knownStructures[s.Name]; ok {
 				return fmt.Errorf("structure name %q is not unique", s.Name)
 			}
-			structureNames[s.Name] = true
+			// keep track of named structures
+			knownStructures[s.Name] = &ps
 		}
+
+		previousEnd = end
 	}
 
-	// validate relative offsets that reference other structures by name
-	for idx, s := range vol.Structure {
-		if s.OffsetWrite.RelativeTo != "" && !structureNames[s.OffsetWrite.RelativeTo] {
-			return fmt.Errorf("structure %v refers to an unknown structure %q",
-				fmtIndexAndName(idx, s.Name), s.OffsetWrite.RelativeTo)
+	// sort by starting offset
+	sort.Sort(byStartOffset(structures))
+
+	return validateCrossVolumeStructure(structures, knownStructures)
+}
+
+func validateCrossVolumeStructure(structures []PositionedStructure, knownStructures map[string]*PositionedStructure) error {
+	previousEnd := Size(0)
+	// cross structure validation:
+	// - relative offsets that reference other structures by name
+	// - positioned structure overlap
+	// use structures positioned within the volume
+	for pidx, ps := range structures {
+		if ps.Role == "mbr" || ps.Type == "mbr" {
+			if ps.StartOffset != 0 {
+				return fmt.Errorf(`structure %v has "mbr" role and must start at offset 0`, ps)
+			}
+		}
+		if ps.OffsetWrite != nil && ps.OffsetWrite.RelativeTo != "" {
+			// offset-write using a named structure
+			other := knownStructures[ps.OffsetWrite.RelativeTo]
+			if other == nil {
+				return fmt.Errorf("structure %v refers to an unknown structure %q",
+					ps, ps.OffsetWrite.RelativeTo)
+			}
+			// offset is written as a 4-byte pointer value at offset-write address
+			if ps.OffsetWrite.Offset > (other.Size - SizeLBA48Pointer) {
```

Comment:
I don't understand this part. Can you expand on the comment please?

Reply:
Pushed a tweak.